flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/rules/physical/stream/StreamExecFirstLastRowRule.scala

```diff
@@ -0,0 +1,180 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.plan.rules.physical.stream
+
+import org.apache.flink.table.calcite.FlinkTypeFactory
+import org.apache.flink.table.plan.`trait`.FlinkRelDistribution
+import org.apache.flink.table.plan.nodes.FlinkConventions
+import org.apache.flink.table.plan.nodes.calcite.ConstantRankRange
+import org.apache.flink.table.plan.nodes.logical.{FlinkLogicalRank, FlinkLogicalSort}
+import org.apache.flink.table.plan.nodes.physical.stream.StreamExecFirstLastRow
+import org.apache.flink.table.plan.util.FlinkRelOptUtil
+
+import org.apache.calcite.plan.{RelOptRule, RelOptRuleCall}
+import org.apache.calcite.rel.`type`.RelDataType
+import org.apache.calcite.rel.convert.ConverterRule
+import org.apache.calcite.rel.core.Sort
+import org.apache.calcite.rel.{RelCollation, RelNode}
+import org.apache.calcite.sql.SqlKind
+
+/**
+  * Rule that matches [[FlinkLogicalSort]] which is sorted by proc-time attribute and
+  * fetches only one record started from 0, and converts it to [[StreamExecFirstLastRow]].
+  */
+class StreamExecFirstLastRowFromSortRule
+  extends ConverterRule(
+    classOf[FlinkLogicalSort],
+    FlinkConventions.LOGICAL,
+    FlinkConventions.STREAM_PHYSICAL,
+    "StreamExecFirstLastRowFromSortRule") {
+
+  override def matches(call: RelOptRuleCall): Boolean = {
+    val sort: FlinkLogicalSort = call.rel(0)
+    StreamExecFirstLastRowRule.canConvertToFirstLastRow(sort)
+  }
+
+  override def convert(rel: RelNode): RelNode = {
+    val sort = rel.asInstanceOf[FlinkLogicalSort]
+    val input = sort.getInput
+    val requiredDistribution = FlinkRelDistribution.SINGLETON
+    val requiredTraitSet = sort.getInput.getTraitSet
+      .replace(FlinkConventions.STREAM_PHYSICAL)
+      .replace(requiredDistribution)
+
+    val convInput: RelNode = RelOptRule.convert(input, requiredTraitSet)
+
+    val fieldCollation = sort.collation.getFieldCollations.get(0)
+    val fieldType = input.getRowType.getFieldList.get(fieldCollation.getFieldIndex).getType
+    val isRowtime = FlinkTypeFactory.isRowtimeIndicatorType(fieldType)
+    val providedTraitSet = rel.getTraitSet.replace(FlinkConventions.STREAM_PHYSICAL)
+    // order by timeIndicator desc ==> lastRow, otherwise is firstRow
+    val isLastRow = fieldCollation.direction.isDescending
+    new StreamExecFirstLastRow(
+      rel.getCluster,
+      providedTraitSet,
+      convInput,
+      Array(),
+      isRowtime,
+      isLastRow)
+  }
+}
+
+/**
+  * Rule that matches [[FlinkLogicalRank]] which is sorted by proc-time attribute and
+  * limits 1 and its RankFunction is ROW_NUMBER, and converts it to [[StreamExecFirstLastRow]].
+  */
+class StreamExecFirstLastRowFromRankRule
```

Comment:
Could you give some examples to show what kind of queries will be captured? And why we are translating these kinds of query to the form you wanted